File: src/test/kotlin/school/charset/app/domain/progress/ModuleProgressTest.kt
package school.charset.app.domain.progress

import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe
import school.charset.app.domain.exercise.ExerciseModule
import kotlin.time.Instant

class ModuleProgressTest :
    FreeSpec({
        val now = Instant.parse("2026-01-01T00:00:00Z")

        // Utf8Encode caps at level 4, so there is headroom to test mid-ladder
        // advancement as well as the ceiling.
        fun progress(level: Int, streak: Int): ModuleProgress = ModuleProgress(
            token = "tok-1",
            module = ExerciseModule.Utf8Encode,
            level = level,
            streak = streak,
            attempts = 0,
            errors = 0,
            lastPlayedAt = null,
        )

        "recordCompletion" - {
            "correct below threshold increments streak, keeps level" {
                val result = progress(level = 1, streak = 3).recordCompletion(correct = true, now = now)

                result.streak shouldBe 4
                result.level shouldBe 1
                result.attempts shouldBe 1
                result.errors shouldBe 0
                result.lastPlayedAt shouldBe now
            }

            "correct hitting threshold advances level and resets streak" {
                // streak 4 + this correct = 5 = STREAK_FOR_LEVEL_UP
                val result = progress(level = 1, streak = 4).recordCompletion(correct = true, now = now)

                result.level shouldBe 2
                result.streak shouldBe 0
                result.attempts shouldBe 1
            }

            "incorrect resets streak and records an error, keeps level" {
                val result = progress(level = 2, streak = 4).recordCompletion(correct = false, now = now)

                result.streak shouldBe 0
                result.level shouldBe 2
                result.errors shouldBe 1
                result.attempts shouldBe 1
            }

            "at max level, hitting the threshold keeps the streak growing (no advance)" {
                // Utf8Encode maxLevel = 4. Already at 4: no further level-up,
                // and the streak is NOT reset so the cap doesn't punish a run.
                val result = progress(level = 4, streak = 4).recordCompletion(correct = true, now = now)

                result.level shouldBe 4
                result.streak shouldBe 5
            }
        }
    })

File: test/charset/progress/module_progress_test.exs
defmodule Charset.Progress.ModuleProgressTest do
  @moduledoc "Pure domain tests - no DB. Port of the Kotlin ModuleProgress specs."
  use ExUnit.Case, async: true

  alias Charset.Progress.ModuleProgress

  @now ~U[2026-06-05 12:00:00Z]

  defp initial, do: ModuleProgress.initial("token-a", :utf8_encode)

  test "initial progression starts at level 1 with empty counters" do
    progress = initial()

    assert progress.level == 1
    assert progress.streak == 0
    assert progress.attempts == 0
    assert progress.errors == 0
    assert progress.last_played_at == nil
  end

  test "a correct completion bumps attempts and streak" do
    progress = ModuleProgress.record_completion(initial(), true, @now)

    assert progress.attempts == 1
    assert progress.errors == 0
    assert progress.streak == 1
    assert progress.last_played_at == @now
  end

  test "an incorrect completion bumps errors and resets the streak" do
    progress =
      initial()
      |> ModuleProgress.record_completion(true, @now)
      |> ModuleProgress.record_completion(true, @now)
      |> ModuleProgress.record_completion(false, @now)

    assert progress.attempts == 3
    assert progress.errors == 1
    assert progress.streak == 0
    assert progress.level == 1
  end

  test "hitting the streak threshold levels up and resets the streak" do
    progress =
      Enum.reduce(1..5, initial(), fn _n, acc ->
        ModuleProgress.record_completion(acc, true, @now)
      end)

    assert progress.level == 2
    assert progress.streak == 0
    assert progress.attempts == 5
  end

  test "the level is capped at the module's max level" do
    # utf16_encode caps at level 2
    progress = %{ModuleProgress.initial("token-a", :utf16_encode) | level: 2, streak: 4}

    progress = ModuleProgress.record_completion(progress, true, @now)

    assert progress.level == 2
    # The streak keeps counting at max level (no reset without a level-up)
    assert progress.streak == 5
  end
end
